torch/_functorch/_aot_autograd/fx_passes.py

```diff
@@ -0,0 +1,202 @@
+from collections import defaultdict
+from typing import Dict, List
+
+import torch
+from torch import fx
+from torch.utils import _pytree as pytree
+
+
+def is_primal(node: fx.Node) -> bool:
+    return node.op == "placeholder" and "tangents" not in str(node.target)
+
+
+def move_resize_zero_to_end_of_graph(graph: fx.Graph) -> None:
+    node_list = list(graph.nodes)
+    return_node = node_list[-1]
+    assert return_node.target == "output"
+    resize_nodes_to_insert_at_end = []
+    node_to_usage_map = defaultdict(list)
+    for i, n in enumerate(node_list):
+        if n.op == "call_function":
+            if (
+                n.target is torch.ops.inductor.resize_storage_bytes_.default
+                and n.args[1] == 0
+            ):
+                resize_nodes_to_insert_at_end.append((n, i))
+            args_flat = pytree.arg_tree_leaves(*n.args, **n.kwargs)
+            for arg in args_flat:
+                node_to_usage_map[arg].append(i)
+    # Assumptions: the size-0 tensor X can then be used only in the following ops:
+    # 1. `aten.set_(X, ...)`
+    # 2. `aten.set_(..., X)`
+    # 3. `inductor.resize_storage_bytes_(X, 0)`
```

Comment:
just curious, but do all 3 of these cases show up in practice? e.g. calling `resize_storage_bytes_(X, 0)` twice in a graph region? Having to deal with multiple resizes on a given tensor definitely complicates the pass a bit, but if we need it we need it

Reply:
Another thing - we do have a few set_ and resize_ tests in `test_aotdispatch.py`, but they mostly just test correctness. We probably want some tests that assert that these passes are properly taking all set_ and resize_ ops in the middle of the graph and properly moving them to the end? (the e2e FSDP test are probably effectively testing this, but the unit tests will probably make these passes easier to reason about over time)

Reply:
> Having to deal with multiple resizes on a given tensor definitely complicates the pass a bit, but if we need it we need it

yeah it's too much complexity for a use case that doesn't exist yet. I just removed all of them :)

Reply:
> but the unit tests will probably make these passes easier to reason about over time

agreed - will add unit tests for this